File: Graphics_Hello_World/java/ppm.java
import java.util.List;
import java.util.ArrayList;
import java.io.IOException;
import java.io.FileWriter;
import java.io.PrintWriter;

class ppm {

    public static void main(String[] args) {
        int width = 256;
        int height = 256;
        ppm image = new ppm();
        image.generateHeader(width, height);
        image.generatePixels(width, height);
        try {
            image.writeImage("javaimage.ppm");
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    String header;
    List<String> pixels;

    public void writeImage(String fileName) throws IOException {
        FileWriter fileWriter = new FileWriter(fileName);
        try(PrintWriter printWriter = new PrintWriter(fileWriter)) {

            printWriter.print(header);
            for (String rgb : pixels) {
                printWriter.print(rgb);
            }
        }
    }

    public void generateHeader(int width, int height) {
      header = new StringBuilder()
          .append("P3\n")
          .append(String.valueOf(width))
          .append(" ")
          .append(String.valueOf(height))
          .append("\n255\n")
          .toString();
    }

    public void generatePixels(int width, int height) {
        pixels = new ArrayList<>(width*height);

        for (int j = height-1; j >= 0; j--) {
            for (int i = 0; i < width; i++) {
                float r = (float) i / (float) width-1;
                float g = (float) j / (float) height-1;
                float b = 0.25f;
                int ir = (int)(255.999f*r);
                int ig = (int)(255.999f*g);
                int ib = (int)(255.999f*b);

                String rgb = new StringBuilder()
                    .append(String.valueOf(ir))
                    .append(" ")
                    .append(String.valueOf(ig))
                    .append(" ")
                    .append(String.valueOf(ib))
                    .append("\n")
                    .toString();

                pixels.add(rgb);
            }
        }
    }
}

File: Graphics_Hello_World/kotlin/ppm.kt
import java.io.File

fun generateHeader(x: Int, y: Int): String {
  val header: String = "P3\n" + x.toString() + " " + y.toString() + "\n255\n"
  return header
}

fun generatePixels(x: Int, y: Int): List<String> {
  val nx: Int = x
  val ny: Int = y
  val pixels = ArrayList<String>()
  for (j in (ny - 1) downTo 0) {
    for (i in 0 until nx) {
      val r: Float = i.toFloat() / nx.toFloat()
      val g: Float = j.toFloat() / ny.toFloat()
      val b: Float = 0.2f
      val ir: Int = (255.99*r).toInt()
      val ig: Int = (255.99*g).toInt()
      val ib: Int = (255.99*b).toInt()
      val rgb: String = ir.toString() + " " + ig.toString() + " " + ib.toString()
      pixels.add(rgb)
    }
  }
  return pixels
}

fun outputImage(fileName: String, header: String, pixels: List<String>){
  File(fileName).printWriter().use { out ->
      out.print(header)
      for (pixel in pixels) {
        out.println(pixel)
      }
  }
}

fun main(args: Array<String>) {
  val x = 200;
  val y = 100;
  val header: String = generateHeader(x, y)
  val pixels = generatePixels(x, y);
  outputImage("kotlinimage.ppm", header, pixels)
}
